Filter and Metafilter code changes for Weaviate Vector DB.

Pragya7011 opened the pull request:
Code changes for adding new parameter of  "filter" and "metafilters" for new search functionality.
Added filter,metafilters parameters with "AND", "OR" and "SIMPLE" operator . 
Also added list of values condition for checking the search statement in list of source,divider,modality and part values as per the user requirement.
Tested below scenarios and getting proper response :

Scenario --

1. List of source files in filter with metafilter conditions
VectorDatabaseQuery ( engine = "cbee987e-1adb-44f7-9196-d4d2b1835245" , command = "What is somatosensory" , limit = 5 , filters = [ Filter ( Source == [ "somatosensory.pdf" , "test.pdf" ] ) ] , metaFilters = [ Filter(Modality == "text")] ) ;

2.Single source file in filter with metafilter conditions
VectorDatabaseQuery ( engine = "cbee987e-1adb-44f7-9196-d4d2b1835245" , command = "What is somatosensory" , limit = 5 , filters = [ Filter ( Source == "somatosensory.pdf" ) ] , metaFilters = [ Filter ( Divider == "1" ) ] ) ;

3.Single source file in filter with metafilter conditions with AND operator
VectorDatabaseQuery ( engine = "cbee987e-1adb-44f7-9196-d4d2b1835245" , command = "What is somatosensory" , limit = 5 , filters = [ Filter ( Source == "somatosensory.pdf" ) ] , metaFilters = [ Filter ( Modality == "text" AND  Divider == "1" ) ] ) ;

4.Single source file in filter with metafilter conditions with OR operator
VectorDatabaseQuery ( engine = "cbee987e-1adb-44f7-9196-d4d2b1835245" , command = "What is somatosensory" , limit = 5 , filters = [ Filter ( Source == "somatosensory.pdf"  OR  Modality == "text" )] ) ;

5.Only search statement : 
VectorDatabaseQuery ( engine = "cbee987e-1adb-44f7-9196-d4d2b1835245" , command = "What is somatosensory" , limit = 5 ) ;

Pragya352 committed:
Filter and Metafilter code changes for Weaviate Vector DB.

themaherkhalil committed:
Merge branch 'dev' into WeaviateFilterChanges

themaherkhalil commented:
The metaFiltersis meant for custom metadata that is added with the document (outside of the base ones we use). 

I don't believe that has been exposed yet for this vector db, but given that all of these values are stored as metadata by definition in weaviate, I see why it works in either situation. 

However, I would expect (and hope when people use this) that everyone uses filters= of metaFilters= for the examples you provided. This way if a person switches from weaviate to pgvector as example, it would end up working since pgvector does have a difference between filters= and metaFilters= as they are stored in different tables and a join is required to work correctly.

themaherkhalil commented:
Could we not change the logic for the filtering to create a master WhereFilter, and then we recursively go and every SimpleQueryFilter will generate a WhereFilter that gets added to the top level construct?  Right now this has a lot of duplicate code and is very hardcoded for the existing columns we have.

Pragya352 committed:
Updated metafilters and code refactoring.

Pragya7011 committed:
Merge branch 'dev' into WeaviateFilterChanges

Pragya7011 commented:
@themaherkhalil  : We have done the code changes for removing duplicate loops and getting WhereFilter response from WeaviateVectorQueryFitler class . Moved the queryfilter logic from WeaviateVectorDatabaseEngine to WeaviateVectorQueryFitler class , removed the unused methods from WeaviateVectorQueryFitler and added constants as part of code refactoring.

Pragya7011 committed:
Merge branch 'dev' into WeaviateFilterChanges

themaherkhalil committed:
Merge branch 'dev' into WeaviateFilterChanges

Subhadeepghosh1 committed:
Merge branch 'dev' into WeaviateFilterChanges

Subhadeepghosh1 committed:
Merge branch 'dev' into WeaviateFilterChanges
